polygeist/tools/cgeist/Lib/clang-mlir.cc

```diff
@@ -4981,28 +5050,26 @@ mlir::func::FuncOp MLIRASTConsumer::GetOrCreateMLIRFunction(
 
   setMLIRFunctionAttributes(function, *FD, lnk, builder.getContext());
 
-  functions[name] = function;
-  module->push_back(function);
-
   if (Def->isThisDeclarationADefinition()) {
     assert(Def->getTemplatedKind() !=
            FunctionDecl::TemplatedKind::TK_FunctionTemplate);
     assert(Def->getTemplatedKind() !=
            FunctionDecl::TemplatedKind::
                TK_DependentFunctionTemplateSpecialization);
     if (ShouldEmit) {
-      functionsToEmit.push_back(Def);
+      functionsToEmit.emplace_back(*Def, Context);
     }
   } else if (ShouldEmit) {
     emitIfFound.insert(name);
   }
-  assert(function->getParentOp() == module.get());
+  checkFunctionParent(function, Context, module, deviceModule);
   return function;
 }
 
 void MLIRASTConsumer::run() {
   while (functionsToEmit.size()) {
-    const FunctionDecl *FD = functionsToEmit.front();
+    auto F = functionsToEmit.front();
```

Comment:
auto -> FunctionToEmit

Reply:
I think in this case this is not needed, as it is quite straightforward that `functionsToEmit` contains `FunctionToEmit`s

Reply:
The name has been chosen carefully, and I agree that in this particular context one can expect that something called `functionsToEmit` might contain elements of type `FunctionsToEmit`. In general the name of a container does not have any relationship with the type of the lements the container has. So I maintain that it is a good programming practice to avoid the use of `auto` unless the type (a) clear because of the context (e.g. a static cast on the RHS of an assignment) or (b) the type is too verbose (e.g. in the case the RHS is a lambda expression).  This is what the LLVM coding guidelines state.